### PR TITLE
Add dedicated server functionality

### PR DESCRIPTION
**Pull request by thearst3rd**

### CODE REVIEW

**Comment by thearst3rd** [comment]
dang conflicts. I'll resolve this next week, after @admoore0's branch gets merged

**Comment by admoore0** [comment]
> dang conflicts. I'll resolve this next week, after @admoore0's branch gets merged

I think these conflicts are just with the mouse sensitivity slider lol, although there will be conflicts with my new branch too.

File: src/autoload/Multiplayer.gd [hunk 1]
```diff
@@ -28,17 +30,68 @@ func _ready():
...
+		get_tree().quit(1)
+		return
+	print("Hosting a dedicated server on port %d" % Global.config.port)
+	get_tree().change_scene("res://src/states/Lobby.tscn")
+
+
+# Attempts to create a server and sets the network peer if successful
+func host_server() -> int:
+	var peer := NetworkedMultiplayerENet.new()
+	var error := peer.create_server(Global.config.port, 4)
```
Comment (admoore0) [inline src/autoload/Multiplayer.gd]: I'd change max_clients to 8, or else have this also be a cmd line arg

File: src/autoload/Multiplayer.gd [hunk 2]
```diff
@@ -127,16 +181,19 @@ remote func register_player(name: String):
 # Disconnect from the session.
 func disconnect_from_session() -> void:
 	Input.mouse_mode = Input.MOUSE_MODE_VISIBLE
-	get_tree().network_peer = null
+	_cleanup_network_peer()
 	player_info = {}
-	var error := get_tree().change_scene("res://src/states/Menu.tscn")
-	assert(not error)
+	if dedicated_server:
```
Comment (admoore0) [inline src/autoload/Multiplayer.gd:208]: Can this even be called from the dedicated server? Isn't this a button action

File: src/autoload/Multiplayer.gd [hunk 3]
```diff
@@ -127,16 +181,19 @@ remote func register_player(name: String):
...
+	else:
+		var error := get_tree().change_scene("res://src/states/Menu.tscn")
+		assert(not error)
 
 
 # Get the player id for this instance. If connected to a server, this is equivalent to the unique
 # network id. If in free play, this will always return 1.
 func get_player_id() -> int:
-	if get_tree().network_peer:
+	if get_tree().has_network_peer():
```
Comment (admoore0) [inline src/autoload/Multiplayer.gd:218]: Big fan of using the setters/getters instead of raw access.

File: src/states/Lobby.gd [hunk 4]
```diff
@@ -70,7 +72,6 @@ func player_connected(player_id: int, info: Dictionary) -> void:
 func player_disconnected(player_id: int) -> void:
 	update_table()
 	if get_tree().is_network_server():
-		# warning-ignore:return_value_discarded
```
Comment (admoore0) [inline src/states/Lobby.gd:83]: Did we change this in editor config?

File: src/states/Lobby.gd [hunk 5]
```diff
@@ -188,13 +188,11 @@ func update_table() -> void:
...
-				all_players_selected = false
-		start_button.disabled = not (all_players_selected and len(chosen_colors.keys()) > 1)
+	if not Multiplayer.dedicated_server:
+		for idx in range(len(COLORS)):
+			button_circle.get_node(str(idx)).disabled = (idx in chosen_colors.values())
+	var all_players_selected := true
+	for player_id in Multiplayer.player_info.keys():
+		if not (player_id in chosen_colors):
+			all_players_selected = false
+	start_button.disabled = not (all_players_selected and len(chosen_colors.keys()) > 0)
```
Comment (admoore0) [inline src/states/Lobby.gd]: Yeah this function is gonna be a huge pain to fix conflicts for lol

**Comment by thearst3rd** [comment]
> > dang conflicts. I'll resolve this next week, after @admoore0's branch gets merged
> 
> I think these conflicts are just with the mouse sensitivity slider lol, although there will be conflicts with my new branch too.

That's why I'm not resolving them now lol

File: src/autoload/Multiplayer.gd (hunk 1 above)
Comment (thearst3rd) [reply src/autoload/Multiplayer.gd]: Good idea, didn't even know that was a variable we had

File: src/autoload/Multiplayer.gd (hunk 2 above)
Comment (thearst3rd) [reply src/autoload/Multiplayer.gd:208]: If someone is running on a desktop platform with a graphical environment, they can still press buttons and such. This is to make those things still work without issues. Someone running using the `Server` platform won't be able to use these ofc

File: src/states/Lobby.gd (hunk 4 above)
Comment (thearst3rd) [reply src/states/Lobby.gd:83]: Yep we did that recently
https://github.com/kdreese/godot-3D-shooter-project/commit/2257cc6c2dfe1633efc258495af44f532100cac5

File: src/states/Lobby.gd (hunk 5 above)
Comment (thearst3rd) [reply src/states/Lobby.gd]: 😩

File: src/autoload/Multiplayer.gd (hunk 1 above)
Comment (admoore0) [reply src/autoload/Multiplayer.gd]: Yeah I think it makes sense to have servers be able to limit the number of clients. Live maybe you want a server just for 1v1s or something

**Review by admoore0 — DISMISSED** [review]

**Review by kdreese — DISMISSED** [review]

**Review by kdreese — APPROVED** [review]

**Review by admoore0 — APPROVED** [review]

**Pull request merged**